templates/fine-tune-llama2/train.py

```diff
@@ -0,0 +1,871 @@
+import argparse
+from filelock import FileLock
+import functools
+import json
+import math
+import os
+import sys
+from pathlib import Path
+import re
+import tempfile
+import time
+import tree
+from typing import Tuple
+
+# check if deepspeed is installed
+try:
+    import deepspeed  # noqa: F401
+except ImportError as e:
+    raise RuntimeError(
+        "Please install deepspeed with `pip install --user deepspeed`."
+    ) from e
+
+from accelerate import Accelerator, DeepSpeedPlugin
+from accelerate.utils import DummyOptim, DummyScheduler, set_seed
+import torch
+import torch.nn as nn
+import tqdm
+from transformers import (
+    AutoModelForCausalLM,
+    AutoTokenizer,
+    get_linear_schedule_with_warmup,
+)
+
+from peft import LoraConfig, get_peft_model
+import ray
+from ray import train
+import ray.util.scheduling_strategies
+from ray.train.torch import TorchTrainer
+from ray.train import Checkpoint
+
+from utils import (
+    get_checkpoint_and_refs_dir,
+    get_mirror_link,
+    download_model,
+    get_download_path,
+)
+
+OPTIM_BETAS = (0.9, 0.999)
+OPTIM_EPS = 1e-8
+NUM_WARMUP_STEPS = 10
+OPTIM_WEIGHT_DECAY = 0.0
+ATTENTION_LAYER_NAME = "self_attn"
+
+
+def get_expected_lora_num_parameters(
```

Comment:
@scottsun94 Please remove this function and its usage.

Reply:
Done